// @react-financial-charts/series/lib/StochasticSeries.js
import * as React from "react";
import { LineSeries } from "./LineSeries";
import { StraightLine } from "./StraightLine";
/**
 * The Stochastic Oscillator is a momentum indicator that shows the location of the close relative to the high-low range over a set number of periods.
 */
export class StochasticSeries extends React.Component {
    constructor() {
        super(...arguments);
        this.yAccessorForK = (d) => {
            const { yAccessor } = this.props;
            return yAccessor(d) && yAccessor(d).K;
        };
        this.yAccessorForD = (d) => {
            const { yAccessor } = this.props;
            return yAccessor(d) && yAccessor(d).D;
        };
    }
    render() {
        const { className, strokeStyle = StochasticSeries.defaultProps.strokeStyle, overSold, middle, overBought, } = this.props;
        return (React.createElement("g", { className: className },
            React.createElement(LineSeries, { yAccessor: this.yAccessorForD, strokeStyle: strokeStyle.dLine }),
            React.createElement(LineSeries, { yAccessor: this.yAccessorForK, strokeStyle: strokeStyle.kLine }),
            React.createElement(StraightLine, { strokeStyle: strokeStyle.top, yValue: overSold }),
            React.createElement(StraightLine, { strokeStyle: strokeStyle.middle, yValue: middle }),
            React.createElement(StraightLine, { strokeStyle: strokeStyle.bottom, yValue: overBought })));
    }
}
StochasticSeries.defaultProps = {
    className: "react-financial-charts-stochastic-series",
    strokeStyle: {
        top: "rgba(150, 75, 0, 0.3)",
        middle: "rgba(0, 0, 0, 0.3)",
        bottom: "rgba(150, 75, 0, 0.3)",
        dLine: "#EA2BFF",
        kLine: "#74D400",
    },
    overSold: 80,
    middle: 50,
    overBought: 20,
};
